File: scripts/backend/elbotola-recovery.py
"""Guarded recovery of the previously approved ElBotola metadata integration.

Only fixed management endpoints and fixed SQL are used. Provider responses,
credentials and command logs never enter the evidence artifact.
"""

from __future__ import annotations

import hashlib
import hmac
import base64
import json
import os
from pathlib import Path
import re
import sys
from datetime import datetime, timezone
from urllib.error import HTTPError, URLError
from urllib.request import HTTPRedirectHandler, Request, build_opener

PROJECT = "tkewgajrljbwgwedqsxn"
PROJECT_NAME = "BotolaGO Production V2"
REPOSITORY = "mrdata007/botolago-foundation"
WORKFLOW = ".github/workflows/news-elbotola-recovery.yml"
MANAGEMENT = f"https://api.supabase.com/v1/projects/{PROJECT}"
FUNCTION_URL = f"https://{PROJECT}.supabase.co/functions/v1/news-ingest-elbotola"
PERMISSION_REFERENCE = "owner-confirmed-2026-08-03-elbotola-link-metadata-and-remote-hero"
COUNTER_NAMES = ("fetched", "validated", "inserted", "updated", "skipped", "rejected", "retries")
FUNCTION_FILES = (
    "supabase/functions/news-ingest-elbotola/index.ts",
    "supabase/functions/_shared/elbotola.ts",
)
VERIFIED_FILES = (WORKFLOW, "scripts/backend/elbotola-recovery.py", *FUNCTION_FILES)

PREFLIGHT_SQL = """
select p.active, p.trust_status::text, p.website_url,
  to_regprocedure('api.news_begin_provider_ingestion(text,text,text)') is not null as has_begin,
  to_regprocedure('api.news_ingest_provider_article(text,text,text,text,text,text,text,text,text,text,text,timestamptz,timestamptz,integer,text)') is not null as has_ingest,
  to_regprocedure('api.news_attach_elbotola_hero(text,text,text)') is not null as has_hero
from app.publishers p where p.slug='elbotola';
"""

ACTIVATE_SQL = """
update app.publishers set active=true, trust_status='trusted', updated_at=statement_timestamp()
where slug='elbotola' and trust_status in ('review_required','trusted')
  and website_url in ('https://www.elbotola.com','https://www.elbotola.com/')
returning active;
"""

VERIFY_SQL = """
select
 (select jsonb_build_object(
   'id',r.id,'status',r.status,'startedAt',r.started_at,
   'fetched',r.records_fetched,'validated',r.records_validated,
   'inserted',r.records_inserted,'updated',r.records_updated,
   'skipped',r.records_skipped,'rejected',r.records_rejected)
  from app_private.news_ingestion_runs r join app.publishers p on p.id=r.publisher_id
  where p.slug='elbotola' order by r.started_at desc limit 1) as run,
 count(*)::integer as published_articles,
 count(*) filter (where e.hero_asset_id is not null)::integer as with_hero,
 max(e.published_at) as latest_publication,
 count(*) filter (where e.language <> 'ar' or e.sanitizer_version <> 'elbotola-link-v1'
  or s.canonical_url !~ '^https://www[.]elbotola[.]com/article/[0-9]{4}-[0-9]{2}-[0-9]{2}-[0-9]{2}-[0-9]{2}-[0-9]+[.]html$')::integer as invalid_metadata
from app.article_editions e join app.stories s on s.id=e.story_id
join app.publishers p on p.id=s.publisher_id
where p.slug='elbotola' and e.status='published' and e.visibility='public' and s.deleted_at is null;
"""


class RecoveryError(Exception):
    pass


class NoRedirect(HTTPRedirectHandler):
    def redirect_request(self, req, fp, code, msg, headers, newurl):
        raise RecoveryError("HTTP_REDIRECT_FORBIDDEN")


def guard(environment):
    expected = environment.get("EXPECTED_COMMIT", "")
    if (
        environment.get("GITHUB_REPOSITORY") != REPOSITORY
        or environment.get("GITHUB_REF") != "refs/heads/main"
        or environment.get("GITHUB_RUN_ATTEMPT") != "1"
        or not re.fullmatch(r"[a-f0-9]{40}", expected)
        or expected != environment.get("GITHUB_SHA")
        or environment.get("CONFIRMATION") != "RUN_ELBOTOLA_RECOVERY"
    ):
        raise RecoveryError("IMMUTABLE_TARGET_GUARD_FAILED")
    event = environment.get("GITHUB_EVENT_NAME")
    mode = environment.get("ELBOTOLA_RECOVERY_MODE")
    if event == "workflow_dispatch":
        if environment.get("GITHUB_ACTOR") != "mrdata007" or mode not in ("canary", "refresh"):
            raise RecoveryError("OPERATOR_GUARD_FAILED")
    elif event == "schedule":
        if mode != "refresh" or environment.get("ELBOTOLA_SCHEDULE_ENABLED") != "true":
            raise RecoveryError("SCHEDULE_NOT_ENABLED")
    else:
        raise RecoveryError("EVENT_NOT_ALLOWED")
    if (
        environment.get("SUPABASE_PRODUCTION_PROJECT_REF") != PROJECT
        or environment.get("SUPABASE_PRODUCTION_PROJECT_NAME") != PROJECT_NAME
        or environment.get("SUPABASE_PRODUCTION_URL", "").rstrip("/") != f"https://{PROJECT}.supabase.co"
    ):
        raise RecoveryError("PRODUCTION_PROJECT_GUARD_FAILED")
    for name in ("SUPABASE_ACCESS_TOKEN", "SUPABASE_SECRET_KEY"):
        value = environment.get(name, "")
        if not value or re.search(r"\s", value):
            raise RecoveryError("PROTECTED_CREDENTIAL_MISSING_OR_INVALID")


def validate_canary_run(run):
    if not isinstance(run, dict) or (
        run.get("path") != WORKFLOW
        or run.get("event") != "workflow_dispatch"
        or run.get("conclusion") != "success"
        or run.get("head_branch") != "main"
        or not re.fullmatch(r"[a-f0-9]{40}", run.get("head_sha", ""))
        or run.get("run_attempt") != 1
        or run.get("repository", {}).get("full_name") != REPOSITORY
    ):
        raise RecoveryError("VERIFIED_CANARY_RUN_REQUIRED")


def dedicated_trigger(secret_key):
    # Domain separation produces an independent credential, reproducible only in
    # the protected runner. It is never a public derivative or the service key.
    return hmac.new(secret_key.encode(), f"botolago:{PROJECT}:elbotola-ingestion:v1".encode(), hashlib.sha256).hexdigest()


def configuration_payload(trigger):
    # No shared GNews or football runtime secret is read, written or deleted.
    values = {
        "ELBOTOLA_SYNDICATION_APPROVED": "true",
        "ELBOTOLA_ORIGIN": "https://www.elbotola.com",
        "ELBOTOLA_PAGE_SIZE": "10",
        "ELBOTOLA_TIMEOUT_MS": "10000",
        "ELBOTOLA_MAX_RETRIES": "1",
        "ELBOTOLA_INGESTION_TRIGGER_SECRET": trigger,
    }
    return [{"name": name, "value": value} for name, value in values.items()]


def validate_response(response):
    if not isinstance(response, dict) or response.get("provider") != "elbotola" or response.get("languages") != ["ar"]:
        raise RecoveryError("UNEXPECTED_INGESTION_RESPONSE")
    counters = response.get("counters")
    if not isinstance(counters, dict) or any(type(counters.get(name)) is not int or counters[name] < 0 for name in COUNTER_NAMES):
        raise RecoveryError("INVALID_INGESTION_COUNTERS")
    if (
        not 1 <= counters["fetched"] <= 10
        or counters["rejected"] != 0
        or counters["validated"] != counters["fetched"]
        or counters["inserted"] + counters["updated"] + counters["skipped"] != counters["fetched"]
    ):
        raise RecoveryError("INGESTION_RECONCILIATION_FAILED")
    return {name: counters[name] for name in COUNTER_NAMES}


def validate_database(result, counters, started):
    if not isinstance(result, list) or len(result) != 1:
        raise RecoveryError("DATABASE_VERIFICATION_FAILED")
    row = result[0]
    run = row.get("run")
    if not isinstance(run, dict) or run.get("status") != "succeeded":
        raise RecoveryError("DATABASE_INGESTION_NOT_SUCCEEDED")
    try:
        run_started = datetime.fromisoformat(run["startedAt"].replace("Z", "+00:00"))
        latest = datetime.fromisoformat(row["latest_publication"].replace("Z", "+00:00"))
    except (KeyError, TypeError, ValueError):
        raise RecoveryError("DATABASE_TIMESTAMPS_INVALID") from None
    if run_started < started or any(run.get(name) != counters[name] for name in COUNTER_NAMES if name != "retries"):
        raise RecoveryError("DATABASE_RUN_DOES_NOT_RECONCILE")
    if row.get("published_articles", 0) < counters["fetched"] or row.get("invalid_metadata") != 0:
        raise RecoveryError("DATABASE_METADATA_INVALID")
    if (started - latest).total_seconds() > 48 * 60 * 60:
        raise RecoveryError("NEWS_STILL_STALE")
    return {name: row[name] for name in ("published_articles", "with_hero", "latest_publication", "invalid_metadata")}


class Recovery:
    def __init__(self, environment):
        guard(environment)
        self.env = environment
        self.directory = Path(environment["ELBOTOLA_RECOVERY_DIR"])
        self.evidence = self.directory / "evidence"
        self.evidence.mkdir(parents=True, mode=0o700, exist_ok=True)
        self.opener = build_opener(NoRedirect())
        self.trigger = dedicated_trigger(environment["SUPABASE_SECRET_KEY"])
        if environment.get("GITHUB_ACTIONS") == "true":
            print("::add-mask::" + self.trigger)

    def request(self, url, token, method="GET", payload=None, headers=None):
        request_headers = {"Authorization": f"Bearer {token}", "Accept": "application/json"}
        request_headers.update(headers or {})
        data = None
        if payload is not None:
            request_headers["Content-Type"] = "application/json"
            data = json.dumps(payload).encode()
        request = Request(url, data=data, headers=request_headers, method=method)
        try:
            with self.opener.open(request, timeout=60) as response:
                raw = response.read(2_000_001)
                if len(raw) > 2_000_000:
                    raise RecoveryError("HTTP_RESPONSE_TOO_LARGE")
                return json.loads(raw) if raw else None
        except HTTPError as error:
            raise RecoveryError(f"HTTP_{error.code}") from None
        except (URLError, TimeoutError, ValueError):
            raise RecoveryError("HTTP_REQUEST_FAILED") from None

    def management(self, path="", method="GET", payload=None):
        if path not in ("", "/secrets", "/database/query", "/functions"):
            raise RecoveryError("MANAGEMENT_PATH_NOT_ALLOWED")
        return self.request(MANAGEMENT + path, self.env["SUPABASE_ACCESS_TOKEN"], method, payload)

    def sql(self, query):
        return self.management("/database/query", "POST", {"query": query})

    def save(self, filename, value):
        target = self.evidence / filename
        target.write_text(json.dumps(value, indent=2) + "\n", encoding="utf-8")
        target.chmod(0o600)

    def state(self):
        return json.loads((self.evidence / "preflight.json").read_text())

    def preflight(self):
        project = self.management()
        if project.get("id") != PROJECT or project.get("name") != PROJECT_NAME or project.get("status") != "ACTIVE_HEALTHY":
            raise RecoveryError("PROJECT_NOT_HEALTHY")
        if self.env["GITHUB_EVENT_NAME"] == "schedule":
            run_id = self.env.get("ELBOTOLA_CANARY_VERIFIED_RUN_ID", "")
            if not re.fullmatch(r"[0-9]+", run_id):
                raise RecoveryError("VERIFIED_CANARY_RUN_REQUIRED")
            run = self.request(f"https://api.github.com/repos/{REPOSITORY}/actions/runs/{run_id}", self.env.get("GITHUB_TOKEN", ""))
            validate_canary_run(run)
            # A manual refresh is not proof of deployment/canary. Verify the two
            # canary-only steps and exact ingestion implementation at that SHA.
            jobs = self.request(f"https://api.github.com/repos/{REPOSITORY}/actions/runs/{run_id}/jobs", self.env.get("GITHUB_TOKEN", ""))
            steps = [step for job in jobs.get("jobs", []) for step in job.get("steps", [])]
            for name in ("Configure isolated ElBotola runtime", "Deploy only reviewed ElBotola function", "Activate publisher and verify bounded ingestion"):
                if not any(step.get("name") == name and step.get("conclusion") == "success" for step in steps):
                    raise RecoveryError("VERIFIED_CANARY_STEPS_REQUIRED")
            for filename in VERIFIED_FILES:
                previous = self.request(f"https://api.github.com/repos/{REPOSITORY}/contents/{filename}?ref={run['head_sha']}", self.env.get("GITHUB_TOKEN", ""))
                if previous.get("encoding") != "base64" or base64.b64decode(previous.get("content", "")) != Path(filename).read_bytes():
                    raise RecoveryError("INGESTION_IMPLEMENTATION_CHANGED_RECANARY_REQUIRED")
        rows = self.sql(PREFLIGHT_SQL)
        if not isinstance(rows, list) or len(rows) != 1:
            raise RecoveryError("ELBOTOLA_PUBLISHER_MISSING")
        publisher = rows[0]
        if publisher.get("trust_status") not in ("trusted", "review_required") or publisher.get("website_url") not in ("https://www.elbotola.com", "https://www.elbotola.com/"):
            raise RecoveryError("ELBOTOLA_PUBLISHER_BLOCKED")
        if any(publisher.get(key) is not True for key in ("has_begin", "has_ingest", "has_hero")):
            raise RecoveryError("ELBOTOLA_DATABASE_PREREQUISITE_MISSING")
        if self.env["ELBOTOLA_RECOVERY_MODE"] == "refresh" and publisher.get("active") is not True:
            raise RecoveryError("ELBOTOLA_PUBLISHER_NOT_ACTIVE")
        self.save("preflight.json", {
            "schemaVersion": 1, "projectRef": PROJECT,
            "expectedCommit": self.env["EXPECTED_COMMIT"], "runId": self.env["GITHUB_RUN_ID"],
            "mode": self.env["ELBOTOLA_RECOVERY_MODE"], "permissionReference": PERMISSION_REFERENCE,
            "previousPublisherActive": publisher["active"], "previousPublisherTrust": publisher["trust_status"],
            "functionSha256": {name: hashlib.sha256(Path(name).read_bytes()).hexdigest() for name in FUNCTION_FILES},
        })

    def configure(self):
        if self.env["ELBOTOLA_RECOVERY_MODE"] != "canary":
            raise RecoveryError("CONFIGURATION_REQUIRES_CANARY")
        self.state()
        payload = configuration_payload(self.trigger)
        self.management("/secrets", "POST", payload)
        self.save("configuration.json", {"configuredNames": [item["name"] for item in payload], "sharedTriggerOverwritten": False})

    def ingest(self):
        state = self.state()
        functions = self.management("/functions")
        matches = [function for function in functions if function.get("slug") == "news-ingest-elbotola"]
        if len(matches) != 1 or matches[0].get("status") != "ACTIVE" or matches[0].get("verify_jwt") is not True:
            raise RecoveryError("ELBOTOLA_SECURE_FUNCTION_MISSING")
        if self.env["ELBOTOLA_RECOVERY_MODE"] == "canary":
            rows = self.sql(ACTIVATE_SQL)
            if len(rows) != 1 or rows[0].get("active") is not True:
                raise RecoveryError("PUBLISHER_ACTIVATION_FAILED")
        started = datetime.now(timezone.utc)
        response = self.request(FUNCTION_URL, self.env["SUPABASE_SECRET_KEY"], "POST", {"job": "elbotola"}, {
            "apikey": self.env["SUPABASE_SECRET_KEY"],
            "x-botolago-ingestion-key": self.trigger,
        })
        counters = validate_response(response)
        database = validate_database(self.sql(VERIFY_SQL), counters, started)
        self.save("result.json", {"schemaVersion": 1, "verdict": "pass", "provider": "elbotola", "languages": ["ar"],
            "expectedCommit": state["expectedCommit"], "runId": self.env["GITHUB_RUN_ID"],
            "mode": state["mode"], "functionVersion": matches[0].get("version"), "counters": counters, "database": database})

    def rollback(self):
        state = self.state()
        # Preserve a previously active integration and all published article history.
        if state["previousPublisherActive"] is False:
            trust = state["previousPublisherTrust"]
            if trust not in ("trusted", "review_required"):
                raise RecoveryError("ROLLBACK_STATE_INVALID")
            self.sql("update app.publishers set active=false, trust_status='" + trust + "', updated_at=statement_timestamp() where slug='elbotola' and trust_status <> 'blocked';")
            self.save("rollback.json", {"publisherRestoredInactive": True, "articlesDeleted": False, "sharedTriggerChanged": False})

    def scan(self):
        values = [self.env.get(name, "") for name in ("SUPABASE_ACCESS_TOKEN", "SUPABASE_SECRET_KEY", "GITHUB_TOKEN")]
        values.append(self.trigger)
        for path in self.evidence.glob("*.json"):
            contents = path.read_text()
            if any(value and value in contents for value in values):
                raise RecoveryError("CREDENTIAL_IN_EVIDENCE")
            if re.search(r"(?:sbp_|sb_secret_|eyJ[A-Za-z0-9_-]+\.)", contents):
                raise RecoveryError("CREDENTIAL_PATTERN_IN_EVIDENCE")


def main():
    recovery = None
    try:
        command = sys.argv[1] if len(sys.argv) == 2 else ""
        if command not in ("preflight", "configure", "ingest", "rollback", "scan"):
            raise RecoveryError("COMMAND_NOT_ALLOWED")
        recovery = Recovery(os.environ)
        getattr(recovery, command)()
        print("ELBOTOLA_" + command.upper() + "_PASS")
        return 0
    except Exception as error:
        code = str(error) if isinstance(error, RecoveryError) else "ELBOTOLA_RECOVERY_FAILED"
        if recovery is not None:
            recovery.save("failure.json", {"verdict": "fail", "code": code})
        print(code)
        return 1


if __name__ == "__main__":
    sys.exit(main())

File: scripts/backend/elbotola-recovery.ts
import { randomBytes } from "node:crypto";
import { mkdir, readFile, writeFile } from "node:fs/promises";
import { resolve } from "node:path";
import { createClient } from "@supabase/supabase-js";
import {
  handleElbotolaRequest,
  type ElbotolaRpcClient,
} from "../../supabase/functions/_shared/elbotola";

type Row = Record<string, unknown>;
const PROJECT = "tkewgajrljbwgwedqsxn";
const REPOSITORY = "mrdata007/botolago-foundation";
const WORKFLOW = ".github/workflows/news-elbotola-recovery.yml";
const VERIFIED_FILES = [
  WORKFLOW,
  "scripts/backend/elbotola-recovery.ts",
  "supabase/functions/_shared/elbotola.ts",
  "supabase/migrations/20260914185233_elbotola_service_source_controls.sql",
  "package.json",
  "bun.lock",
];
const COUNTERS = [
  "fetched",
  "validated",
  "inserted",
  "updated",
  "skipped",
  "rejected",
  "retries",
] as const;
type Counters = Record<(typeof COUNTERS)[number], number>;
const UUID = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;

export class ElbotolaRecoveryError extends Error {
  constructor(readonly code: string) {
    super(code);
  }
}
function fail(code: string): never {
  throw new ElbotolaRecoveryError(code);
}
function row(value: unknown): Row {
  if (!value || typeof value !== "object" || Array.isArray(value)) fail("invalid_recovery_payload");
  return value as Row;
}

export function runtimeGuard(env: NodeJS.ProcessEnv) {
  if (
    env.GITHUB_REPOSITORY !== REPOSITORY ||
    env.GITHUB_REF !== "refs/heads/main" ||
    env.GITHUB_RUN_ATTEMPT !== "1" ||
    !/^[a-f0-9]{40}$/.test(env.EXPECTED_COMMIT ?? "") ||
    env.EXPECTED_COMMIT !== env.GITHUB_SHA ||
    env.CONFIRMATION !== "RUN_ELBOTOLA_RECOVERY"
  )
    fail("immutable_target_guard_failed");
  const mode = env.ELBOTOLA_RECOVERY_MODE;
  if (env.GITHUB_EVENT_NAME === "workflow_dispatch") {
    if (env.GITHUB_ACTOR !== "mrdata007" || (mode !== "canary" && mode !== "refresh"))
      fail("owner_dispatch_required");
  } else if (
    env.GITHUB_EVENT_NAME !== "schedule" ||
    mode !== "refresh" ||
    env.ELBOTOLA_SCHEDULE_ENABLED !== "true"
  ) {
    fail("schedule_not_enabled");
  }
  const url = env.SUPABASE_PRODUCTION_URL?.replace(/\/$/, "");
  if (
    env.SUPABASE_PRODUCTION_PROJECT_REF !== PROJECT ||
    env.SUPABASE_PRODUCTION_PROJECT_NAME !== "BotolaGO Production V2" ||
    url !== `https://${PROJECT}.supabase.co`
  )
    fail("production_target_mismatch");
  if (!env.SUPABASE_SECRET_KEY || /\s/.test(env.SUPABASE_SECRET_KEY))
    fail("protected_service_key_missing");
  if (
    !env.SUPABASE_PRODUCTION_PUBLISHABLE_KEY?.startsWith("sb_publishable_") ||
    /\s/.test(env.SUPABASE_PRODUCTION_PUBLISHABLE_KEY)
  )
    fail("protected_publishable_key_missing");
  return {
    mode: mode as "canary" | "refresh",
    commit: env.EXPECTED_COMMIT!,
    url,
    key: env.SUPABASE_SECRET_KEY,
    publishableKey: env.SUPABASE_PRODUCTION_PUBLISHABLE_KEY,
  };
}

export function createRecoveryClients(
  config: ReturnType<typeof runtimeGuard>,
  fetcher?: typeof fetch,
) {
  const options = {
    auth: { persistSession: false, autoRefreshToken: false, detectSessionInUrl: false },
    global: fetcher ? { fetch: fetcher } : undefined,
  };
  return {
    writer: createClient(config.url, config.key, options),
    reader: createClient(config.url, config.publishableKey, options),
  };
}

export function validateCanaryRun(value: unknown): string {
  const run = row(value);
  if (
    run.path !== WORKFLOW ||
    run.event !== "workflow_dispatch" ||
    run.conclusion !== "success" ||
    run.head_branch !== "main" ||
    run.run_attempt !== 1 ||
    row(run.repository ?? {}).full_name !== REPOSITORY ||
    row(run.actor ?? {}).login !== "mrdata007" ||
    typeof run.head_sha !== "string" ||
    !/^[a-f0-9]{40}$/.test(run.head_sha)
  )
    fail("verified_manual_canary_required");
  return run.head_sha;
}

async function verifyPriorCanary(env: NodeJS.ProcessEnv): Promise<Row> {
  const runId = env.ELBOTOLA_CANARY_VERIFIED_RUN_ID ?? "";
  if (!/^[1-9]\d*$/.test(runId) || !env.GITHUB_TOKEN) fail("verified_manual_canary_required");
  const request = async (path: string): Promise<Row> => {
    const response = await fetch(`https://api.github.com/repos/${REPOSITORY}/${path}`, {
      headers: {
        Accept: "application/vnd.github+json",
        Authorization: `Bearer ${env.GITHUB_TOKEN}`,
        "X-GitHub-Api-Version": "2022-11-28",
      },
      redirect: "error",
      signal: AbortSignal.timeout(30_000),
    });
    if (!response.ok) fail("canary_evidence_unavailable");
    const text = await response.text();
    if (text.length > 2_000_000) fail("canary_evidence_too_large");
    return row(JSON.parse(text));
  };
  const commit = validateCanaryRun(await request(`actions/runs/${runId}`));
  const jobs = await request(`actions/runs/${runId}/jobs?per_page=100`);
  if (
    !Array.isArray(jobs.jobs) ||
    !jobs.jobs.some((job) => {
      const steps = row(job).steps;
      return (
        Array.isArray(steps) &&
        steps.some(
          (step) =>
            row(step).name === "Confirm manual ElBotola canary completed" &&
            row(step).conclusion === "success",
        )
      );
    })
  )
    fail("verified_manual_canary_steps_required");
  for (const filename of VERIFIED_FILES) {
    const previous = await request(`contents/${filename}?ref=${commit}`);
    if (
      previous.encoding !== "base64" ||
      typeof previous.content !== "string" ||
      !Buffer.from(previous.content, "base64").equals(await readFile(filename))
    )
      fail("ingestion_implementation_changed_recanary_required");
  }
  return { runId, commit, implementationUnchanged: true };
}

export function validateCounters(value: unknown): Counters {
  const response = row(value);
  if (response.provider !== "elbotola" || JSON.stringify(response.languages) !== '["ar"]')
    fail("unexpected_ingestion_response");
  const counters = row(response.counters);
  for (const name of COUNTERS) {
    if (
      typeof counters[name] !== "number" ||
      !Number.isSafeInteger(counters[name]) ||
      counters[name] < 0
    )
      fail("invalid_ingestion_counters");
  }
  const result = counters as Counters;
  if (
    result.fetched < 1 ||
    result.fetched > 10 ||
    result.rejected !== 0 ||
    result.validated !== result.fetched ||
    result.inserted + result.updated + result.skipped !== result.fetched
  )
    fail("ingestion_reconciliation_failed");
  return result;
}

export interface ObservedArticle {
  id: string;
  canonicalUrl: string;
  publishedAt: string;
}
/**
 * BG-0073 News stand-down.
 *
 * Ingestion no longer publishes: api.news_ingest_provider_article lands every
 * edition as draft/private, so an ingested article is deliberately NOT reachable
 * through the public feed. This check is therefore the inverse of what it used
 * to be - it proves the stand-down holds, that a scheduled or dispatched run
 * still cannot put third-party link-outs in front of a reader.
 *
 * The freshness gate is kept, but it is now measured against the provider
 * publication timestamps we actually observed being persisted rather than
 * against the public feed, which is empty by design. That is the check's real
 * purpose: proving the provider is handing us current content.
 */
export function validateStandDownFeed(value: unknown, articles: ObservedArticle[], now: Date): Row {
  const feed = row(value);
  if (
    !Array.isArray(feed.items) ||
    articles.length < 1 ||
    new Set(articles.map((article) => article.id)).size !== articles.length
  )
    fail("public_feed_reconciliation_failed");
  const items = feed.items.map(row);
  const ingested = new Set(articles.map((article) => article.id));
  if (items.some((item) => typeof item.id === "string" && ingested.has(item.id)))
    fail("ingested_article_public_after_stand_down");
  let latest = 0;
  for (const article of articles) {
    const published = Date.parse(article.publishedAt);
    if (!Number.isFinite(published)) fail("public_feed_reconciliation_failed");
    latest = Math.max(latest, published);
  }
  if (
    !Number.isFinite(latest) ||
    latest < now.getTime() - 48 * 60 * 60 * 1_000 ||
    latest > now.getTime() + 5 * 60 * 1_000
  )
    fail("news_still_stale");
  return {
    ingestedArticles: articles.length,
    publicFeedItems: items.length,
    reachableAfterStandDown: 0,
    latestIngestedPublication: new Date(latest).toISOString(),
  };
}

export function observedClient(
  client: ElbotolaRpcClient,
  evidence: Row,
  articles: ObservedArticle[],
): ElbotolaRpcClient {
  return {
    schema: () => ({
      rpc: async (name, args) => {
        const result = await client.schema("api").rpc(name, args);
        if (result.error) return result;
        if (name === "news_begin_provider_ingestion") {
          const runId = row(result.data).runId;
          if (typeof runId !== "string" || !UUID.test(runId)) fail("invalid_ingestion_run_id");
          evidence.ingestionRunId = runId;
        } else if (name === "news_ingest_provider_article") {
          const id = row(result.data).articleId;
          if (typeof id !== "string" || !UUID.test(id)) fail("invalid_ingested_article_id");
          articles.push({
            id,
            canonicalUrl: String(args.p_canonical_url),
            publishedAt: String(args.p_source_published_at),
          });
        } else if (name === "news_complete_ingestion_run") {
          evidence.databaseCompletion = {
            status: args.p_status,
            runId: args.p_run_id,
            counters: Object.fromEntries(
              COUNTERS.filter((name) => name !== "retries").map((name) => [
                name,
                args[`p_${name}`],
              ]),
            ),
          };
        }
        return result;
      },
    }),
  };
}

export async function prepareSource(
  client: ElbotolaRpcClient,
  mode: "canary" | "refresh",
): Promise<boolean> {
  const status = await client.schema("api").rpc("service_elbotola_source_status", {});
  if (status.error) fail("elbotola_source_status_unavailable");
  const source = row(status.data);
  if (
    typeof source.active !== "boolean" ||
    !["trusted", "review_required"].includes(String(source.trustStatus)) ||
    !["https://www.elbotola.com", "https://www.elbotola.com/"].includes(String(source.websiteUrl))
  )
    fail("elbotola_source_blocked");
  if (mode === "refresh" && !source.active) fail("elbotola_source_not_active");
  return source.active;
}

async function main(): Promise<void> {
  const config = runtimeGuard(process.env);
  const directory = resolve(
    process.env.ELBOTOLA_RECOVERY_DIR ?? fail("missing_evidence_directory"),
    "evidence",
  );
  await mkdir(directory, { recursive: true, mode: 0o700 });
  const trigger = randomBytes(32).toString("hex");
  if (process.env.GITHUB_ACTIONS === "true") console.log(`::add-mask::${trigger}`);
  const evidence: Row = {
    schemaVersion: 2,
    expectedCommit: config.commit,
    projectRef: PROJECT,
    runId: process.env.GITHUB_RUN_ID,
    mode: config.mode,
    provider: "elbotola",
    languages: ["ar"],
    permissionReference: "owner-confirmed-2026-08-03-link-metadata-and-remote-hero",
    verdict: "in_progress",
  };
  const { writer: database, reader: publicDatabase } = createRecoveryClients(config);
  let previousActive: boolean | null = null;
  let activatedFromInactive = false;
  const save = async () => {
    const text = JSON.stringify(evidence, null, 2) + "\n";
    if (
      [config.key, config.publishableKey, trigger, process.env.GITHUB_TOKEN].some(
        (secret) => secret && text.includes(secret),
      )
    )
      fail("credential_in_evidence");
    await writeFile(resolve(directory, "result.json"), text, { mode: 0o600 });
  };
  try {
    if (config.mode === "refresh") evidence.verifiedCanary = await verifyPriorCanary(process.env);
    previousActive = await prepareSource(database as unknown as ElbotolaRpcClient, config.mode);
    evidence.previousPublisherActive = previousActive;
    if (config.mode === "canary" && !previousActive) {
      const activation = await database
        .schema("api")
        .rpc("service_set_elbotola_source_active", { p_active: true });
      if (
        activation.error ||
        typeof row(activation.data).previousActive !== "boolean" ||
        row(activation.data).currentActive !== true
      )
        fail("elbotola_source_activation_failed");
      previousActive = row(activation.data).previousActive as boolean;
      activatedFromInactive = previousActive === false;
      evidence.previousPublisherActive = previousActive;
    }
    const articles: ObservedArticle[] = [];
    const client = observedClient(database as unknown as ElbotolaRpcClient, evidence, articles);
    const response = await handleElbotolaRequest(
      new Request(`${config.url}/functions/v1/news-ingest-elbotola`, {
        method: "POST",
        headers: { "content-type": "application/json", "x-botolago-ingestion-key": trigger },
        body: '{"job":"elbotola"}',
      }),
      {
        client,
        environment: {
          ELBOTOLA_INGESTION_TRIGGER_SECRET: trigger,
          ELBOTOLA_SYNDICATION_APPROVED: "true",
          ELBOTOLA_ORIGIN: "https://www.elbotola.com",
          ELBOTOLA_PAGE_SIZE: "10",
          ELBOTOLA_TIMEOUT_MS: "10000",
          ELBOTOLA_MAX_RETRIES: "1",
        },
      },
    );
    const result = row(await response.json());
    if (!response.ok)
      fail(
        typeof result.error === "string" && /^[a-z_]{3,80}$/.test(result.error)
          ? result.error
          : "ingestion_handler_failed",
      );
    const counters = validateCounters(result);
    evidence.counters = counters;
    const completion = row(evidence.databaseCompletion);
    if (
      completion.status !== "succeeded" ||
      completion.runId !== evidence.ingestionRunId ||
      articles.length !== counters.fetched ||
      COUNTERS.some(
        (name) => name !== "retries" && row(completion.counters)[name] !== counters[name],
      )
    )
      fail("database_completion_mismatch");
    const feed = await publicDatabase
      .schema("api")
      .rpc("news_feed", { p_language: "ar", p_limit: 50 });
    if (feed.error) fail("public_feed_unavailable");
    evidence.publicFeed = validateStandDownFeed(feed.data, articles, new Date());
    // BG-0073: an ingested article is draft/private, so api.news_article_detail
    // deliberately refuses to serve it. Confirm that refusal rather than the
    // outbound link, which no public route can show while News is stood down.
    const detail = await publicDatabase
      .schema("api")
      .rpc("news_article_detail", { p_language: "ar", p_identifier: articles[0].id });
    if (!detail.error && row(detail.data ?? {}).id === articles[0].id)
      fail("ingested_article_readable_after_stand_down");
    evidence.verdict = "pass";
    await save();
    console.log("ELBOTOLA_RECOVERY_PASS");
  } catch (error) {
    evidence.verdict = "fail";
    evidence.failureCode =
      error instanceof ElbotolaRecoveryError ? error.code : "elbotola_recovery_failed";
    if (config.mode === "canary" && activatedFromInactive) {
      try {
        const rollback = await database
          .schema("api")
          .rpc("service_set_elbotola_source_active", { p_active: false });
        evidence.publisherRestoredInactive =
          !rollback.error && row(rollback.data).currentActive === false;
      } catch {
        evidence.publisherRestoredInactive = false;
      }
    }
    await save();
    throw new ElbotolaRecoveryError(String(evidence.failureCode));
  }
}

if (import.meta.main)
  main().catch((error) => {
    console.error(error instanceof ElbotolaRecoveryError ? error.code : "elbotola_recovery_failed");
    process.exitCode = 1;
  });
